Vertex AI entrypoint (#750)
zenml-io/zenml: docs/book/advanced-guide/getting_started_on_gcp.md

# Getting started with Google Cloud

> This guide will take you from the default local stack to a basic cloud 

Reviewer: Ok the only thing missing now IMO is the motivation of this section. We need to make clear what this guide is about. We can say something like "To get started using ZenML on the cloud, you need some basic infrastructure up and running that you can then make more complicated depending on your use-case. This guide sets up the easiest MLOPs stack that we can run on GCP with ZenML" or osmething?

P.S. I would call the title of this doc only `GCP` for simplicity

stack in no time. This should help you get started with running your machine 
learning pipeline in production. 

{% hint style="info" %}
This guide represents **one** of many ways to create a cloud stack on GCP. 
Every component could be replaced by a different implementation.
{% endhint %}

## Prerequisites 

For this to work you need to have zenml installed locally with all GCP 
requirements.

```bash
pip install zenml
zenml integration install gcp
```

Additionally, you will need Docker  installed on your system. 

## The cloud stack

A full cloud stack will necessarily contain these five stack components:

* An **artifact store** to save all step output artifacts, in this guide we will
use a GCP bucket for this purpose
* A **metadata store** that keeps track of the relationships between artifacts, 
runs and parameters. In our case we will opt for a MySQL database on GCP
Cloud SQL.
* The **orchestrator** to run the pipelines. Here we will opt for a Vertex AI
pipelines orchestrator. This is a serverless GCP specific offering with minimal
hassle.
* A **container registry** for pushing and pulling the pipeline image.

> * Finally, the **Secrets Manager** to store passwords and SSL certificates.

Reviewer 2: ```suggestion
* Finally, the **secrets manager** to store passwords and SSL certificates.
```
For consistency across the names of stack components.

## Step 1/10 Set up a GCP project (Optional)

As a first step it might make sense to 
[create](https://console.cloud.google.com/projectcreate) 

> a separate GCP project for your zenml resources. However, this step is 

Reviewer 2: ```suggestion
a separate GCP project for your ZenML resources. However, this step is 
```

completely optional, and you can also move forward within an existing project. 

> If some resources already exist, feel free to skip the creation of the 

Reviewer 2: ```suggestion
If some resources already exist, feel free to skip their creation step
```

> resources and simply note down the relevant information. 

Reviewer 2: ```suggestion
and simply note down the relevant information. 
```

For simplicity, just
open up a terminal on the side and export relevant values as we go along. You
will use these when we set up the ZenML stack.
ZenML will use your project number at a later stage to connect to some 
resources, so let's export it. You'll most probably find it right 
[here](https://console.cloud.google.com/welcome).

```bash
export PROJECT_NUMBER=<PROJECT_NUMBER> # for example '492014921912'
```

## Step 2/10 Enable billing

Before moving on, you'll have to make sure you attach a billing account to 
your project. In case you do not have the permissions to do so, you'll have to

> ask an organization administrator to do so. 

Reviewer 2: ```suggestion
ask an organization administrator.
```

> [Here](https://console.cloud.google.com/billing/projects) is the relevant 

Reviewer 2: ```suggestion
[Here](https://console.cloud.google.com/billing/projects) is a relevant page.
```

> page to do so.

Reviewer 2: ```suggestion

```

## Step 3/10 Enable Vertex AI

Vertex AI pipelines is at the heart of our GCP stack. As the orchestrator 
Vertex AI will run your pipelines and use all the other stack components. 
All you'll need to do at this stage is enable Vertex AI
[here](https://console.cloud.google.com/vertex-ai).

Once it is enabled you will see a drop-down with all the regions where 
Vertex AI is available. At this point it might make sense to make this 
decision for your ZenML Stack and export the full region name.

```shell
export GCP_LOCATION=<GCP_LOCATION> # for example 'europe-west3'
```

> ## Step 4/10 Enable SSL ger

Reviewer 2: ```suggestion
## Step 4/10 Enable Secrets Manager
```

> The SSL ger will be needed so that the orchestrator will have secure

Reviewer 2: ```suggestion
The secrets manager will be needed so that the orchestrator will have secure
```

access to the other resources. 
[Here](https://console.cloud.google.com/marketplace/product/google/secretmanager.googleapis.com)
is where you'll be able to enable the secrets manager.

## Step 5/10 Enable Container Registry

The Vertex AI orchestrator uses Docker Images containing your pipeline code
for pipeline orchestration. For this to work you'll need to enable the GCP
Docker  registry 
[here](https://console.cloud.google.com/marketplace/product/google/containerregistry.googleapis.com).

In order to use the container registry at a later point you will need to 

> set the container registry uri. This is how it is usually constructed:

Reviewer 2: ```suggestion
set the container registry URI. This is how it is usually constructed:
```

`gcr.io/<PROJECT_ID>`. 

{% hint style="info" %}

> The container registry has four options: **gcr**.io , us.gcr.io , eu.gcr.io , 

Reviewer 3: Does it look better if we format this into `gcr.io` , `us.gcr.io` , `eu.gcr.io`?

or `asia.gcr.io`. Choose the one appropriate for you. 
{% endhint %}

```bash
export CONTAINER_REGISTRY_URI=<CONTAINER_REGISTRY_URI> # for example 'eu.gcr.io/zenml-project'
```

## Step 6/10 Set up Cloud Storage as Artifact Store

Storing of step artifacts is an important part of reproducible MLOps. 
Create a bucket [here](https://console.cloud.google.com/storage/create-bucket).

Within the configuration of the newly created bucket you can find the 

> gsutil URI which you will need at a later point. Its usually going to look like 

Reviewer 2: ```suggestion
gsutil URI which you will need at a later point. It's usually going to look like 
```

this: `gs://<bucket-name>`

```bash
export GSUTIL_URI=<GSUTIL_URI> # for example 'gs://zenml_vertex_storage'
```

## Step 7/10 Set up a Cloud SQL instance as Metadata Store

One of the most complex resources that you'll need to manage is the MySQL
database. 

To start, we [create](https://console.cloud.google.com/sql/instances/create;engine=MySQLe)
a MySQL database. Once created, it will take some time for the database to be 
set up. 

Once it is set up you can find the IP-address. The password you set during 
creation of the instance is the root password. The default port for MySQL is 
3306. 

```bash
export DB_HOST_IP=<DB_HOST_IP> # for example '35.137.24.15'
export DB_PORT=<DB_PORT> # usually by default '3306'
export DB_USER=<DB_USER> # 'root' if you don't set up a separate user
export DB_PWD=<DB_PWD> # for example 'secure_root_pwd'
```

Time to set up the connections to our database. To do this you'll need to go 
into the `Connections` menu. Under the `Networking` tab you'll need to add 

> **0.0.0.0** to the authorized networks, thereby allowing all incoming traffic 

Reviewer 2: ```suggestion
**0.0.0.0/0** to the authorized networks, thereby allowing all incoming traffic 
```

from everywhere. (Feel free to restrict this to your outgoing IP address)

For security reasons, it is also recommended to configure your database to only 

> accept SSL connections. You'll find the relevant setting in the **Security*+ 

Reviewer 2: ```suggestion
accept SSL connections. You'll find the relevant setting in the **Security** 
```

export SSL_CERT=@<SSL_CERT> # for example @/home/zen/Downloads/client-cert.pem
export SSL_KEY=@<SSL_KEY> # for example @/home/zen/Downloads/client-key.pem
```

Finally, head on over to the `Databases` submenu and create your database and
export its name. 

```bash
export DB_NAME=<DB_NAME> # for example zenml_db
```

## Step 8/10 - Create a Service Account

All the resources are created. Now we need to make sure the instance performing 
the compute engine (Vertex AI) needs to have the relevant permissions to access 
the other resources. For this you'll need to go
[here](https://console.cloud.google.com/iam-admin/serviceaccounts/create) to 
create a service account. Give it a relevant name and allow access to the
following roles:

* Vertex AI Custom Code Service Agent
* Vertex AI Service Agent
* Container Registry Service Agent
* Secret Manager Admin

Also give your user access to the service account. This is the service account 
that will be used by the Vertex AI compute engine.

```bash
export SERVICE_ACCOUNT=<SERVICE_ACCOUNT> # for example zenml-vertex-sa@zenml-project.iam.gserviceaccount.com
```

On top of this we will also need to give permissions to the service account

> of the custom code workers. For this head over to your IAM 

Reviewer 2: ```suggestion
of the custom code workers. 

For this, head over to your IAM 
```

[configurations](https://console.cloud.google.com/iam-admin/iam), click on 
**Include Google-provided role grants** on the top right and find the 
**<project_number>@gcp-sa-aiplatform-cc.iam.gserviceaccount.com** service
account. 

Now give this one the **Container Registry Service Agent** role.

> ## Step 9/10 Set Up gcloud CLI

Reviewer 2: ```suggestion
## Step 9/10 Set Up `gcloud` CLI
```

> Install the gcloud cli on your machine. 

Reviewer 2: ```suggestion
Install the `gcloud` CLI on your machine. 
```

┠────────────────────┼─────────────────────┨
┃ CONTAINER_REGISTRY │ gcp_registry        ┃
┠────────────────────┼─────────────────────┨
┃ METADATA_STORE     │ gcp_metadata_store  ┃
┠────────────────────┼─────────────────────┨
┃ ORCHESTRATOR       │ vertex_orchestrator ┃
┠────────────────────┼─────────────────────┨
┃ SECRETS_MANAGER    │ gcp_secrets_manager ┃
┗━━━━━━━━━━━━━━━━━━━━┷━━━━━━━━━━━━━━━━━━━━━┛
     'gcp_vertex_stack' stack (ACTIVE)  
```

## Run your pipeline in the cloud

With your ZenML stack set up and active, you are now ready to run your ZenML
pipeline on Vertex AI. 

For example, you could pull the ZenML Vertex AI example and run it.

```bash
zenml example pull vertex_ai_orchestration
cd zenml_examples/vertex_ai_orchestration/
python run.py
```

{% hint style="info" %}
Your first run might fail as one of the service accounts is only created once 
vertex is run for the first time. This service account will need to be given
appropriate rights after the first run fails.
{% endhint %}

At the end of the logs you should be seeing a link to the Vertex AI dashboard. 
It should look something like this:

![Finished Run](../assets/VertexAiRun.png)

> In case you get an error message like this:

Reviewer 2: Is there a possibility that this error doesn't occur for a new user? If not, then we can directly include the following instructions as part of creating the service account as a subsection to avoid the chance of running into this error in the first place.

Reviewer 2: Oh I see, the service account isn't available before the first run. That is super weird

pipelines orchestrator. This is a serverless GCP specific offering with minimal
hassle.
* A **container registry** for pushing and pulling the pipeline image.

> * Finally, the **SSL ger** to store passwords and SSL certificates.

Reviewer 3: Unintended typo SSL ger?
